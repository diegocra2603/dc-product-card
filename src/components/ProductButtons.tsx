import React , { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
import { ProductCardExtendStylesProps } from '../interfaces';

export interface Props extends ProductCardExtendStylesProps {
}

export const ProductButtons = ({ className, style }: Props) => {

    const { increaseBy, counter, maxCount } = useContext(ProductContext);

    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount,
        [counter],
    )

    const isMinReached = useCallback(
        () => counter === 0,
        [counter],
    )

    const HandlerClickIncrease = (value: number) => {

        if (value === +1 && !isMaxReached()) {
            increaseBy(value)
        }

        if (value === -1 && !isMinReached()) {
            increaseBy(value)
        }
    }

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={`${styles.buttonMinus} ${isMinReached() && styles.disable}`} onClick={() => HandlerClickIncrease(-1)} >-</button>
            <span className={styles.countLabel}>{counter}</span>
            <button className={`${styles.buttonAdd} ${isMaxReached() && styles.disable}`} onClick={() => HandlerClickIncrease(+1)}>+</button>
        </div>
    )
}