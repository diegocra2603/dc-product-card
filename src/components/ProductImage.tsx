import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';

import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductCardExtendStylesProps } from '../interfaces';

export interface Props extends ProductCardExtendStylesProps {
    img?: string;
}

export const ProductImage = ({ img, className, style }: Props) => {

    const { product } = useContext(ProductContext);

    let imgToShow: string;

    if (img) {
        imgToShow = img;
    } else if (product.img) {
        imgToShow = product.img
    } else {
        imgToShow = noImage
    }

    return (
        <img className={`${styles.productImg} ${className}`} src={imgToShow} alt="Product Image" style={style} />
    )
}