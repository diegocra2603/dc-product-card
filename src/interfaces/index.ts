import { Props as ProductButtonsProps } from "../components/ProductButtons";
import { Props as ProductCardProps } from "../components/ProductCard";
import { Props as ProductImageProps } from "../components/ProductImage";
import { Props as ProductTitleProps } from "../components/ProductTitle";

export interface Product {
    id: string,
    title: string;
    img?: string;
}

export interface ProductContextProps {
    counter: number;
    increaseBy: (value: number) => void;
    product: Product;
    maxCount: number;
}

export interface ProductCardExtendStylesProps {
    className?: string;
    style?: React.CSSProperties
}

export interface onChangeArgs {
    product: Product,
    count: number
}

export interface ProductInCart extends Product {
    count: number
}

export interface InitialValues {
    count?: number;
    maxCount?: number;
}

export interface ProductCardHandlers {
    count: number;
    isMaxCountReached: boolean;
    maxCount?: number;
    product: Product;

    increaseBy: (value: number) => void;
    reset: () => void;
}

export interface ProductCardHOCProps {
    ({ }: ProductCardProps): JSX.Element;
    Title: ({ }: ProductTitleProps) => JSX.Element;
    Image: ({ }: ProductImageProps) => JSX.Element;
    Buttons: ({ }: ProductButtonsProps) => JSX.Element;
}