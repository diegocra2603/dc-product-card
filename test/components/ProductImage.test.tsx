import React from "react";
import render from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from "../data/products";


describe('ProductImage', () => {

    test('debe de mostrar el componente correctamente con la imagen personalizada', () => {
        const wrapper = render.create(
            <ProductImage className="custom-img" img="./coffee-mug.png" />
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    })

    test('debe de mostar el component con la imagen del producto', () => {
        const wrapper = render.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })

});