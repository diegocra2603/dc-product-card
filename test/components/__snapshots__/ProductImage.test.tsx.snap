// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostar el component con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product Image"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage debe de mostrar el componente correctamente con la imagen personalizada 1`] = `
<img
  alt="Product Image"
  className="productImg custom-img"
  src="./coffee-mug.png"
/>
`;
